### PR TITLE
Change subscribeToSaveResponse to use finally for cleaning isSaving.

### PR DESCRIPTION
**Pull request by mshima**
---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by kaidohallik** [comment]
This will not work. According to http://reactivex.io/documentation/contract.html `onComplete` is not executed if error occurs.

**Comment by mshima** [comment]
Switched to use finalize instead of onComplete

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 1]
```diff
@@ -103,6 +103,22 @@ describe('Component Tests', () => {
                     expect(comp.isSaving).toEqual(false);
                 })
             );
+
+            it('Should cleanup on error',
+                fakeAsync(() => {
+                    // GIVEN
+                    const entity = new <%= entityAngularName %>(<%- tsKeyId %>);
+                    spyOn(service, 'update').and.returnValue(throwError('This is an error!'));
+                    comp.updateForm(entity);
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:112]: This should be `ngOnInit()` and data should come from `ActivatedRoute`.

Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:112]: ActivatedRoute Mocking is not available on this branch.
We should switch to it on others tests too.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 2]
```diff
@@ -30,7 +30,7 @@ import { HttpResponse } from '@angular/common/http';
 import { HttpClientTestingModule } from '@angular/common/http/testing';
 import { FormBuilder } from '@angular/forms';
 import { ActivatedRoute } from '@angular/router';
-import { of } from 'rxjs';
+import { of, throwError } from 'rxjs';
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:33]:
```suggestion
import { of, Subject } from 'rxjs';
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 3]
```diff
@@ -200,6 +200,24 @@ describe('Component Tests', () => {
...
+
+                    // WHEN
+                    comp.save();
+                    tick(); // simulate async
+
+                    // THEN
+                    expect(<%= entityInstance %>Service.update).toHaveBeenCalledWith(<%= entityInstance %>);
+                    expect(comp.isSaving).toEqual(false);
+                })
+            );
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:220]:
```suggestion
            it('Should set isSaving to false on error', () => {
                // GIVEN
                const saveSubject = new Subject();
                const <%= entityInstance %> = new <%= entityAngularName %>(<%- tsKeyId %>);
                spyOn(<%= entityInstance %>Service, 'update').and.returnValue(saveSubject);
                activatedRoute.data = of({ <%= entityInstance %> });
                comp.ngOnInit();

                // WHEN
                comp.save();
                expect(comp.isSaving).toEqual(true);
                saveSubject.error('This is an error!');

                // THEN
                expect(<%= entityInstance %>Service.update).toHaveBeenCalledWith(<%= entityInstance %>);
                expect(comp.isSaving).toEqual(false);
            });
```

**Review by kaidohallik — CHANGES_REQUESTED** [review]
As `isSaving` logic changed (previously this was set to `false` on `next`, now on `complete`, this is ok because `HttpClient` methods complete after emitting value) then it would be good to show this in tests in all variants of `save` method.
First 2 tests of save method could be:
```
            it('Should call update service on save for existing entity', () => {
                // GIVEN
                const saveSubject = new Subject();
                const <%= entityInstance %> = new <%= entityAngularName %>(<%- tsKeyId %>);
                spyOn(<%= entityInstance %>Service, 'update').and.returnValue(saveSubject);
                activatedRoute.data = of({ <%= entityInstance %> });
                comp.ngOnInit();

                // WHEN
                comp.save();
                expect(comp.isSaving).toEqual(true);
                saveSubject.next(new HttpResponse({ body: <%= entityInstance %> }));
                saveSubject.complete();

                // THEN
                expect(<%= entityInstance %>Service.update).toHaveBeenCalledWith(<%= entityInstance %>);
                expect(comp.isSaving).toEqual(false);
            });

            it('Should call create service on save for new entity', () => {
                // GIVEN
                const saveSubject = new Subject();
                const <%= entityInstance %> = new <%= entityAngularName %>();
                spyOn(<%= entityInstance %>Service, 'create').and.returnValue(saveSubject);
                activatedRoute.data = of({ <%= entityInstance %> });
                comp.ngOnInit();

                // WHEN
                comp.save();
                expect(comp.isSaving).toEqual(true);
                saveSubject.next(new HttpResponse({ body: <%= entityInstance %> }));
                saveSubject.complete();

                // THEN
                expect(<%= entityInstance %>Service.create).toHaveBeenCalledWith(<%= entityInstance %>);
                expect(comp.isSaving).toEqual(false);
            });
```

**Comment by kaidohallik** [comment]
One more idea: it would be good to test also that successful saving returns to previous page and unsuccessful stays in the same page. For example add to every `save` test `GIVEN` part:
```
        spyOn(comp, 'previousState');
```
and in tests for successful completion add to `THEN` part:
```
        expect(comp.previousState).toHaveBeenCalled();
```
and in test for unsuccessful completion add to `THEN` part:
```
        expect(comp.previousState).not.toHaveBeenCalled();
```

**Comment by kaidohallik** [comment]
`fakeAsync` and `tick` imports are now redundant
https://github.com/jhipster/generator-jhipster/pull/13776/checks?check_run_id=1815044090#step:27:116

**Comment by mshima** [comment]
@kaidohallik fell free to edit my PRs and commit.

**Review by kaidohallik — APPROVED** [review]

**Pull request merged**